docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,122 @@
+title = "SIP 017 - Service Chaining"
+template = "main"
+date = "2024-02-21T23:00:00Z"
+
+---
+
+Summary: Allow in-process HTTP requests to application components without going via the network
+
+Owner(s): ivan.towlson@fermyon.com
+
+Created: Feb 21, 2024
+
+# Requirements and scope
+
+In Spin 2.2, a component can make a HTTP request with only a path (e.g. `/api/accounts`).  Such a request will be routed to the same application, but this is done by Spin prepending the host where the application is running (e.g. `https://mysrsbizness.fermyon.app/api/accounts`).  The consequence of this is that requests go out to the network and come back in, including through load balancers etc.  This adds overhead, and in some environnments could count against quota or incur bandwidth charges.
+
+_Service chaining_ refers to carrying out HTTP requests in memory, without going off the node, by invoking the receiving Spin component directly instead of via the HTTP server.
+
+We have [prototyped](https://github.com/fermyon/spin/pull/2229) dispatching HTTP requests directly to the component handling the route, but this was in the context of an implementation detail of Spin.  The requirement for this project is for an application to declare that it _requires_ dispatch to run locally on the node, and for hosts that cannot meet that prerequisite to reject the application.  This allows developers to write applications as a network of chained microservices, without having to be concerned about what happens if the microservices are not co-located.
+
+# User experience
+
+## Expressing a requirement for chaining
+
+To express that the application requires service chaining, add the following to `spin.toml`:
+
+```toml
+[application.host_requirements]
+spin_internal_requests = "required"
+```
+
+> Why ask the user to write out `"required"` instead of just listing the names?  In future, we may have features where it makes sense to say "I'd prefer to run on a host that supports this, but it's not a blocker," or where the user can specify a value (for a bad example, consider an "available memory" prerequisite).
```

Comment:
Decision: we will not surface this in the TOML, but instead infer it from the presence of `spin.internal` in `allowed_outbound_hosts`.  It will still appear in an up-level lock file.